### PR TITLE
LRQA-64845 | master

### PR DESCRIPTION
**Pull request by Tim-Cao**
https://issues.liferay.com/browse/LRQA-64845

This is only for master.

@david-gutierrez-mesa
@john-co

### CODE REVIEW

**Comment by Tim-Cao** [comment]
ci:test:sf

**Comment by david-gutierrez-mesa** [comment]
LGTM!

@john-co can you check it as well?

**Comment by ealonso** [comment]
ci:test:relevant

**Comment by john-co** [comment]
imo, having the padding hardcoded and specific like this will be hard to maintain in long run. Unless we have some design requirements/references to maintain the padding long term? I don't know how often the padding values will change but if they're not expected to change frequently then this is ok short to mid term.

If not have requirements...some other things to explore to make it more flexible, is to set the padding to a certain tolerable expected range. Something like check the values are all are greater than 0px? or compare the value is > 10px and <20px? 

Also, I'd recommend setting the value as a constant global property for product menu. Something like 
```
var minProductMenuCategoryRightPadding = 10px
var minProductMenuCategoryLeftPadding = 10px
```

**Comment by Tim-Cao** [comment]
ci:test:sf

**Comment by Tim-Cao** [comment]
ci:test:relevant

**Comment by Tim-Cao** [comment]
@john-co, I've modified according to your advice. Could you check the case again?
https://github.com/liferay-echo/liferay-portal/pull/3817/commits/d4f2a69ca104d7aa174daa74b054c00b6137f9f3

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/wem/productmenu/ProductMenu.testcase [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+		task ("Assert the padding settings of category") {
+			AssertVisible(
+				key_category = "Design",
+				locator1 = "ProductMenu#CATEGORY_COLLAPSED");
+
+			for (var paddingVertical : list "padding-bottom,padding-top") {
+				TestUtils.assertCssValueIsGreaterThan(
+					attribute = "${paddingVertical}",
+					element = "//a[@data-qa-id='appGroup'][normalize-space(text())='Design']",
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/wem/productmenu/ProductMenu.testcase:47]: Could we use a `.path` locator name, or this only works with literal xpath?

**Review by john-co — COMMENTED** [review]
Looks good overall, I added some comments to organize better.

File: portal-web/test/functional/com/liferay/portalweb/macros/TestCase.macro [hunk 2]
```diff
@@ -17,6 +17,25 @@ definition {
...
+
+return computedStyle.getPropertyValue("${attribute}");
+
+		''';
+
+		var cssValue = selenium.getEval("${javascript}");
+
+		return "${cssValue}";
+	}
+
```
[the remark this answers is not in the document]
Comment (Tim-Cao) [reply portal-web/test/functional/com/liferay/portalweb/macros/TestCase.macro:38]: The existing macro in Css.macro is required to update. The document.querySelector using id attribute value is less reusable than document.evaluate(locator, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;

File: portal-web/test/functional/com/liferay/portalweb/tests/development/tools/testinginfrastructure/block/macro/TestUtils.macro [hunk 3]
```diff
@@ -1,5 +1,16 @@
...
+		var actual = TestCase.getCssValue(
+			attribute = "${attribute}",
+			element = "${element}");
+		var actual = StringUtil.extractLeadingDigits("${actual}");
+
+		TestUtils.isGreaterThan(
+			actual = "${actual}",
+			expected = "${expected}");
+	}
+
```
[the remark this answers is not in the document]
Comment (Tim-Cao) [reply portal-web/test/functional/com/liferay/portalweb/tests/development/tools/testinginfrastructure/block/macro/TestUtils.macro:13]: I will move it to Css.macro

File: portal-web/test/functional/com/liferay/portalweb/macros/TestCase.macro (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tim-Cao) [reply portal-web/test/functional/com/liferay/portalweb/macros/TestCase.macro:38]: I will modify the existing macro.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/wem/productmenu/ProductMenu.testcase (hunk 1 above)
Comment (Tim-Cao) [reply portal-web/test/functional/com/liferay/portalweb/tests/enduser/wem/productmenu/ProductMenu.testcase:47]: There are two limitation in Poshi. One is function layer cannot return value. The other one is only function layer can invoke locator layer but macro cannot.

Comment (john-co) [reply portal-web/test/functional/com/liferay/portalweb/tests/enduser/wem/productmenu/ProductMenu.testcase:47]: makes sense, ok.

**Comment by Tim-Cao** [comment]
ci:test:sf

**Comment by Tim-Cao** [comment]
ci:test:sf

**Comment by john-co** [comment]
changes lgtm! ok to forward.

**Comment by Tim-Cao** [comment]
ci:forward

**Pull request closed without merging**